### PR TITLE
Improve TypeScript definitions

### PR DESCRIPTION
**Pull request by BendingBender**
Fix return type of `p-throttle` to signal that the throttled function always returns a `Promise` and that there is a `cancel` method on the returned function.

### CODE REVIEW

**Comment by BendingBender** [comment]
@sindresorhus Could you please review the changes?

File: index.d.ts [hunk 1]
```diff
@@ -25,4 +27,17 @@ export class AbortError extends Error {
  * 	throttled(i).then(console.log);
  * }
  */
-export default function <T extends Function>(input: T, limit: number, interval: number): T;
+export default function<TArguments extends any[], TReturn>(
+	fn: (...arguments: TArguments) => PromiseLike<TReturn> | TReturn,
```
Comment (sindresorhus) [inline index.d.ts]: `fn` => `function`

File: index.d.ts [hunk 2]
```diff
@@ -25,4 +27,17 @@ export class AbortError extends Error {
  * 	throttled(i).then(console.log);
  * }
  */
-export default function <T extends Function>(input: T, limit: number, interval: number): T;
+export default function<TArguments extends any[], TReturn>(
```
Comment (sindresorhus) [inline index.d.ts]: I prefer not prefixing types with `T`.

`any` => `unknown`

File: index.d.ts [hunk 3]
```diff
@@ -25,4 +27,17 @@ export class AbortError extends Error {
...
+	fn: (...arguments: TArguments) => PromiseLike<TReturn> | TReturn,
+	limit: number,
+	interval: number
+): ThrottledFunction<TArguments, TReturn>;
+
+export type ThrottledFunction<TArguments extends any[], TReturn> = ((
+	...args: TArguments
```
Comment (sindresorhus) [inline index.d.ts]: `args` => `arguments`, for consistency.

File: index.d.ts [hunk 4]
```diff
@@ -25,4 +27,17 @@ export class AbortError extends Error {
...
  * }
  */
-export default function <T extends Function>(input: T, limit: number, interval: number): T;
+export default function<TArguments extends any[], TReturn>(
+	fn: (...arguments: TArguments) => PromiseLike<TReturn> | TReturn,
+	limit: number,
+	interval: number
+): ThrottledFunction<TArguments, TReturn>;
+
+export type ThrottledFunction<TArguments extends any[], TReturn> = ((
```
Comment (sindresorhus) [inline index.d.ts]: Move the type above the function.

File: index.d.ts [hunk 5]
```diff
@@ -1,14 +1,16 @@
 export class AbortError extends Error {
+	readonly name: 'AbortError';
```
Comment (sindresorhus) [inline index.d.ts]: Why is this needed? The name is already defined in the `class AbortError`.

File: index.test-d.ts [hunk 6]
```diff
@@ -1,15 +1,26 @@
 import {expectType} from 'tsd-check';
-import pThrottle, {AbortError} from '.';
+import pThrottle, {AbortError, ThrottledFunction} from '.';
 
-const throttledUnicorn = pThrottle((index: string) => {
-	return '🦄';
-}, 1, 1000);
+const throttledUnicorn = pThrottle(
+	(index: string) => {
+		return '🦄';
```
Comment (sindresorhus) [inline index.test-d.ts]: Just use a inline arrow function.

File: index.d.ts (hunk 5 above)
Comment (BendingBender) [reply index.d.ts]: It can be useful if you try to find out the type of the error without using `typeof`:

```ts
class E extends Error {
    name: 'E';
    foo: 'bar';
}

const e = new E();

if (e.name === 'E') {
    e.foo; // <- no error here
}
```

Comment (BendingBender) [reply index.d.ts]: I suppose for the sake of completeness and for the case that you use this error in a discriminated union type:

```ts
class E extends Error {
    name: 'E';
    foo: 'bar';
}

class F extends Error {
    name: 'F';
    baz: 'bar';
}

var e : E | F = new E();

if (e.name === 'E') {
    e.foo; // <- no TS error here
}
```

File: index.d.ts (hunk 1 above)
Comment (BendingBender) [reply index.d.ts]: This doesn't work, TS seems to forbid the `function` a keyword here.

File: index.d.ts (hunk 2 above)
Comment (BendingBender) [reply index.d.ts]: > I prefer not prefixing types with `T`.

I'll change this, but don't you think that this actually helps distinguishing between actual types and type variables? This is such a common pattern in TS, if I'd read the types file, I'd automatically assume (without further verifying) that `Arguments` must be some declared type in the module, but `TArguments` must be a type variable.

File: index.d.ts (hunk 5 above)
Comment (sindresorhus) [reply index.d.ts]: Makes sense, but I'm surprised TS does not make the `.name` property the same as the name of the class...

Comment (BendingBender) [reply index.d.ts]: Well, this is not the class property that is being set here (`AbortError.name`) but the instance property name (`err.name`). The first is being set automatically, the second not.

**Comment by sindresorhus** [comment]
Would you consider this a breaking change for the package? I'm not sure whether it should be a minor or major version bump.

**Comment by BendingBender** [comment]
Well, technically following was done:
- fixes for a broken type definition (the return type)
- improve types in a way that they are more specific (correct argument type checks)
- `unknown` type is now used, which raises the min TS version to 3.0

These changes will certainly produce errors for incorrect usages of `p-throttle` and for the correct ones that use TS < 3.0. We didn't rename any exported types here, so at least this seems OK.

DefinitelyTyped has given up on this issue, they don't consider incompatible improvements major changes but this is due to the fact that they want to align typedef versions with lib versions.

If it was my lib, I'd make it a major release.

**Comment by sindresorhus** [comment]
Thanks for the insight. I agree, it's better to do a major version than risk breaking someone's build.
